docs/report_stage_i.md

```diff
@@ -0,0 +1,497 @@
+For the preliminary analysis, the OPCODE `COINBASE` was chosen arbitrarily, this choice will be revisited.
+
+We'll use implementation-relative measurements whenever we want to compare the relative proportions of OPCODE measurements between environments (mainly different implementations).
+
+#### Estimation validation and exploration
+
+Here we outline the choice of options for validations to do on the obtained measurements.
+
+1. **compare OPCODE measurement statistics between themselves and between different implementations** - inspect the means/quantiles/variances/distributions and look for abnormalities, list out and attempt to explain them.
+1. **compare individual instruction vs entire program execution** - see how well does the former predict the latter (see [Individual measurement vs entire program execution measurement](#Individual-measurement-vs-entire-program-execution-measurement))
+2. **analyze individual OPCODEs dynamics** - generate a multitude of programs whereby the OPCODE is ran in varying circumstances. See if the measurement error is random and we're accounting for all the dynamics of computational cost of the OPCODE.
+3. **cross-environment OPCODE validation** - check OPCODEs which have drastically different relative estimates in different implementations/hardwares, look into whether the reason is intrinsic to the implementation/hardware or is due to error.
+4. **impact on gas cost schedule** - estimate the gas costs, calibrate to the excluded OPCODEs and compare with current gas schedule. Repeat for estimations from various environments and methods. Analyze the impact of gas cost adjustments.
+3. **historical validation** - check against blockchain history under normal conditions where the node is ran (see [Instrumenting and measuring the computations from blockchain history](#Instrumenting-and-measuring-the-computations-from-blockchain-history))
+4. **cross-timer validation** - capture all results using an alternative CPU cost proxy (e.g. instead of `runtimeNano` use `gotsc` CPU cycles), see how they compare.
+
+#### Instrumenting and measuring the computations from blockchain history
+
+For an additional validation of the gas cost estimates, we consider instrumenting and measuring a fragment of the blockchain.
+
+The problem arises from the fact that our gas cost estimator is prepared only for a subset of instructions (in particular, no IO/storage instructions).
+We will consider calibrating these together by picking a "pivot OPCODE", whose gas cost would remain unchanged, and adjusting gas cost for all other OPCODE from the analyzed subset to match relative difference in cost to the pivot OPCODE.
```

Comment:
It is not clear how picking the pivot opcode should solve the problem with IO/storage instructions.

Reply:
@mbenke the problem is that we don't estimate for IO/storage instructions at all. So if we want to validate using contracts which touch IO/storage, we should relate our gas cost estimations (for a subset of OPCODEs), with the rest.

Say we pick `COINBASE` as pivot (arbitrary choice right now). all other OPCODEs from our subset are expressed as multiples of `COINBASE`. Since currently `COINBASE` costs ~1~ 2 gas, we derive new gas cost values for our subset of OPCODEs. The remainder (IO/storage etc) remains unadjusted.

Does this make sense? Should I clarify in text using this explanation?

Reply:
@pdobacz Yes, I see what you mean. Perhaps it's a question of language again. It wouyld be grat if you could rephrase these two sentences for clarity.